README.md
# dialog

now you can talk to your computer in simple yet wonderful ways, look at that!

## dependencies

you're gonna need the haskell tool `stack` to build this guy, if/when you got
that you can clone this repo and `cd` to it and run

```
stack build
```

to build it and then you can run

```
stack exec prompt <command> <prompt>
```

to test it from that same project directory and

```
stack install
```

to install it so you can run

```
prompt <command> <prompt>
```

from whichever directory your heart desires

you're also gonna need `sdl2`, `sdl2-ttf`, and `sdl2-gfx`. so also install
those by whatever means you prefer according to your computer and operating
system and all that jazz.
 
## font

a path to a font is hardcoded in `src/PromptWindow.hs` so you will prolly have
to change this path to point to a font that is on your system to get
things to work. the relevant line of code looks like:

```
font <- F.load "/usr/share/fonts/TTF/Crimson-Roman.ttf" 42
```


CHANGELOG.md
# Changelog for `gui-thing`

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to the
[Haskell Package Versioning Policy](https://pvp.haskell.org/).

## Unreleased

## 0.1.0.0 - YYYY-MM-DD


src/Lib.hs
{-# LANGUAGE ImportQualifiedPost, OverloadedStrings #-}

module Lib (windowPrompt) where

import Foreign.C.Types (CInt)
import Data.Text       (Text, pack, singleton)
import Control.Monad   (mapM)

import SDL
import SDL.Font      qualified as F
import SDL.Primitive qualified as P

import Selection (Selection(..), toggle)
import Waves (sinwaveI, coswaveI, ellipsewaveV2, rectwaveV2)

width, height :: CInt
(width, height) = (640, 400)

-- Position
type Pos = V2 CInt
-- Dimension
type Dim = Pos

windowPrompt :: String -> IO Bool
windowPrompt question = do
  initializeAll
  F.initialize

  window <- createWindow (pack $ "prompt : " ++ question) defaultWindow { windowBorder = False, windowInitialSize = V2 width height }
  renderer <- createRenderer window (-1) defaultRenderer
  font <- F.load "/usr/share/fonts/TTF/Crimson-Roman.ttf" 42

  response <- promptLoop renderer font question Y 0

  destroyWindow window
  F.quit
  quit

  pure response


renderPrompt :: (Integral a) 
             => Renderer 
             -> F.Font 
             -> String
             -> a
             -> IO ()
renderPrompt renderer font prompt i =
  do surfs <- mapM ((F.blended font (V4 0 0 255 255)) . singleton) prompt
     dims  <- mapM surfaceDimensions surfs

     let bounds = map (\d -> (V2 (width - 9) (height `div` 3)) - d) dims
         points = map (\(b, j) -> (V2 0 (height `div` 6)) + (ellipsewaveV2 b 500 (i + 250 + (j * 6)))) 
                      (zip bounds [0..])
         boxes = map (\(p, d) -> (Rectangle (P p) d))
                     (zip points dims)

     textures <- mapM (createTextureFromSurface renderer) surfs

     mapM_ (\(t,b) -> copy renderer t Nothing (Just b)) (zip textures boxes)
     -- box demarcating ellipse boundaries
     P.rectangle renderer (V2 0 (height `div` 6))
                          (V2 (width - 9) ((height `div` 6) + (height `div` 3))) 
                          (V4 0 0 255 255)


     mapM_ freeSurface surfs
     mapM_ destroyTexture textures

renderY :: (Integral a)
        => Renderer
        -> F.Font
        -> Selection
        -> a
        -> IO Dim
renderY renderer font sel i =
  do surf <- F.blended font (V4 0 255 127 (1 + (sinwaveI 255 200 i))) "y"
                                        -- ^-- 0 alpha means opaque, so add 1
     dim  <- surfaceDimensions surf

     let pos = V2 (width   `div` 4) 
                  ((height `div` 4) * 3)
         box = (Rectangle (P pos) dim)

     tex <- createTextureFromSurface renderer surf

     copy renderer tex Nothing (Just box)

     freeSurface surf
     destroyTexture tex

     pure dim

renderN :: (Integral a)
        => Renderer
        -> F.Font
        -> Selection
        -> a
        -> IO Dim
renderN renderer font sel i =
  do surf <- F.blended font (V4 255 0 127 (1 + (sinwaveI 255 200 (i + 100)))) "n"
                                        -- ^-- 0 alpha means opaque, so add 1
     dim  <- surfaceDimensions surf

     let pos = V2 ((width  `div` 4) * 3)
                  ((height `div` 4) * 3)
         box = (Rectangle (P pos) dim)

     tex <- createTextureFromSurface renderer surf

     copy renderer tex Nothing (Just box)

     freeSurface surf
     destroyTexture tex

     pure dim

renderSelection :: (Integral a)
                => Renderer
                -> Selection
                -> Dim
                -> Dim
                -> a
                -> IO ()
renderSelection renderer sel yDim nDim i =
  do let margins = (pure 5) :: Dim
         padding = (V2 5 0) :: Dim
     case sel of
       Y -> let pos = V2 (width   `div` 4)
                         ((height `div` 4) * 3)
                box1Cycle = rectwaveV2 margins 50 i
                box2Cycle = rectwaveV2 margins 50 (i + 25)
             in do P.rectangle renderer (pos + box1Cycle - padding - margins) (pos + box1Cycle + yDim + padding) (V4 0 0 255 255)
                   P.rectangle renderer (pos + box2Cycle - padding - margins) (pos + box2Cycle + yDim + padding) (V4 0 255 127 255)
       N -> let pos = V2 ((width  `div` 4) * 3)
                         ((height `div` 4) * 3)
                box1Cycle = rectwaveV2 margins 50 i
                box2Cycle = rectwaveV2 margins 50 (i + 25)
             in do P.rectangle renderer (pos + box1Cycle - margins - padding) (pos + box1Cycle + nDim + padding) (V4 0 0 255 255)
                   P.rectangle renderer (pos + box2Cycle - margins - padding) (pos + box2Cycle + nDim + padding) (V4 255 0 127 255)


promptLoop :: Renderer  -- renderer to display things onto
           -> F.Font    -- font to display with
           -> String    -- the prompt to display
           -> Selection -- option (either y or n) currently selected
           -> Int       -- iterator
           -> IO Bool   -- true if y is selected, false if n is
promptLoop renderer font prompt sel i = do
  rendererDrawColor renderer $= V4 0 0 0 255
  clear renderer

  renderPrompt renderer font prompt i
  yDim <- renderY renderer font sel i
  nDim <- renderN renderer font sel i
  renderSelection renderer sel yDim nDim i

  present renderer

  delay 25

  events <- pollEvents
  let isKeyPress keycode event =
        case eventPayload event of
          KeyboardEvent kbe ->
            keyboardEventKeyMotion kbe == Pressed &&
            keysymKeycode (keyboardEventKeysym kbe) == keycode
          _ -> False
      yPressed     = any (isKeyPress KeycodeY) events
      nPressed     = any (isKeyPress KeycodeN) events
      tabPressed   = any (isKeyPress KeycodeTab) events
      enterPressed = any (isKeyPress KeycodeReturn) events

  -- TODO: add mouse support (click y/n to do what y/n does)
  --       also change selection on hover (i.e. if mouse hovers 
  --       over y, move selection to y)
  -- y means do the command given as argument
  -- n means don't do that command
  if yPressed || enterPressed && sel == Y then pure True
  else if nPressed || enterPressed && sel == N then pure False
  else if tabPressed then (promptLoop renderer font prompt (toggle sel) (i + 1))
  else (promptLoop renderer font prompt sel (i + 1))


src/Waves.hs
module Waves (sinwaveI 
             ,coswaveI
             ,ellipsewaveV2
             ,rectwaveV2) where

import SDL.Vect (V2, V2(..))

-- I means Integral, it means the function yeilds an Integral
-- maybe I'll make variants with F for floating some time
waveI :: (RealFloat a, Integral b, Integral c, Integral d)
     => (a -> a) -- a trig function (or any function that will map Z to [-1,1]
                 --                  over a period of length 2pi
     -> b        -- amplitude
     -> b        -- wavelength
     -> c        -- curent position in cycle (the 'x' value)
     -> d        -- current position on band (the 'y' value)
waveI f amplitude wavelength =
  let toRadians  = (((2 * pi) / (fromIntegral wavelength)) *)
      recenter   = (/ 2) . (1 +)
      factorBand = ((fromIntegral amplitude) *)
   in round . factorBand . recenter . f . toRadians . fromIntegral

sinwaveI :: (Integral a, Integral b, Integral c) => a -> a -> b -> c
sinwaveI = waveI sin

coswaveI :: (Integral a, Integral b, Integral c) => a -> a -> b -> c
coswaveI = waveI cos

ellipsewaveV2 :: (Integral a, Integral b)
       => (V2 a) -- width/height to stay within
       -> a      -- length of cycle (larger for slower movement)
       -> b      -- current position in cycle
       -> (V2 a) -- current position in ellipse
ellipsewaveV2 box wavelength i =
  let w = (\(V2 x _) -> coswaveI x wavelength i) box
      h = (\(V2 _ y) -> sinwaveI y wavelength i) box
   in (V2 w h)

rectwaveV2 :: (Integral a, Integral b)
       => (V2 a) -- width/height to stay within
       -> a      -- length of cycle (larger for slower movement)
       -> b      -- current position in cycle
       -> (V2 a) -- current position in rectangle type thing
rectwaveV2 box wavelength i =
  let squarish_coswave = waveI (sin . (* pi) . (/ 2) . cos)
      squarish_sinwave = waveI (sin . (* pi) . (/ 2) . sin)
      w = (\(V2 x _) -> squarish_coswave x wavelength i) box
      h = (\(V2 _ y) -> squarish_sinwave y wavelength i) box
   in (V2 w h)



src/PromptWindow.hs
{-# LANGUAGE ImportQualifiedPost, OverloadedStrings #-}

module PromptWindow (promptWindow) where

import Foreign.C.Types (CInt)
import Data.Text       (Text, pack, singleton)
import Control.Monad   (mapM)

import SDL
import SDL.Font      qualified as F
import SDL.Primitive qualified as P

import Selection (Selection(..), toggle)
import Waves (sinwaveI, coswaveI, ellipsewaveV2, rectwaveV2)

width, height :: CInt
(width, height) = (640, 400)

-- Position
type Pos = V2 CInt
-- Dimension
type Dim = Pos

promptWindow :: String -> IO Bool
promptWindow question =
  do initializeAll
     F.initialize

     window <- createWindow (pack $ "prompt : " ++ question) defaultWindow { windowBorder = False
                                                                           , windowInitialSize = V2 width height }
     renderer <- createRenderer window (-1) defaultRenderer
     font <- F.load "/usr/share/fonts/TTF/Crimson-Roman.ttf" 42

     response <- promptLoop renderer font question Y 0

     destroyWindow window
     F.quit
     quit

     pure response


renderPrompt :: (Integral a) 
             => Renderer 
             -> F.Font 
             -> String
             -> a
             -> IO ()
renderPrompt renderer font prompt i =
  do surfs <- mapM ((F.blended font (V4 0 0 255 255)) . singleton) prompt
     dims  <- mapM surfaceDimensions surfs

     let bounds = map (\d -> (V2 (width - 9) (height `div` 3)) - d) dims
         points = map (\(b, j) -> (V2 0 (height `div` 6)) + (ellipsewaveV2 b 500 (i + 250 + (j * 6)))) 
                      (zip bounds [0..])
         rects = map (\(p, d) -> (Rectangle (P p) d))
                     (zip points dims)

     textures <- mapM (createTextureFromSurface renderer) surfs

     mapM_ (\(t,b) -> copy renderer t Nothing (Just b)) (zip textures rects)
     -- box demarcating ellipse boundaries
     P.rectangle renderer (V2 0 (height `div` 6))
                          (V2 (width - 9) ((height `div` 6) + (height `div` 3))) 
                          (V4 0 0 255 255)


     mapM_ freeSurface surfs
     mapM_ destroyTexture textures

renderY :: (Integral a)
        => Renderer
        -> F.Font
        -> a
        -> IO Dim
renderY renderer font i =
  do surf <- F.blended font (V4 0 255 127 (1 + (sinwaveI 255 200 i))) "y"
                                        -- ^-- 0 alpha means opaque, so add 1
     dim  <- surfaceDimensions surf

     let pos = V2 (width   `div` 4) 
                  ((height `div` 4) * 3)
         box = (Rectangle (P pos) dim)

     tex <- createTextureFromSurface renderer surf

     copy renderer tex Nothing (Just box)

     freeSurface surf
     destroyTexture tex

     pure dim

renderN :: (Integral a)
        => Renderer
        -> F.Font
        -> a
        -> IO Dim
renderN renderer font i =
  do surf <- F.blended font (V4 255 0 127 (1 + (sinwaveI 255 200 (i + 100)))) "n"
                                        -- ^-- 0 alpha means opaque, so add 1
     dim  <- surfaceDimensions surf

     let pos = V2 ((width  `div` 4) * 3)
                  ((height `div` 4) * 3)
         box = (Rectangle (P pos) dim)

     tex <- createTextureFromSurface renderer surf

     copy renderer tex Nothing (Just box)

     freeSurface surf
     destroyTexture tex

     pure dim

renderSelection :: (Integral a)
                => Renderer
                -> Selection
                -> Dim
                -> Dim
                -> a
                -> IO ()
renderSelection renderer sel yDim nDim i =
  let margins = (pure 5) :: Dim
      padding = (V2 5 0) :: Dim
   in case sel of
        Y -> let pos = V2 (width   `div` 4)
                         ((height `div` 4) * 3)
                 box1Cycle = rectwaveV2 margins 40 i
                 box2Cycle = rectwaveV2 margins 40 (i + 20)
              in do P.rectangle renderer (pos + box1Cycle - padding - margins) (pos + box1Cycle + yDim + padding) (V4 0 0 255 255)
                    P.rectangle renderer (pos + box2Cycle - padding - margins) (pos + box2Cycle + yDim + padding) (V4 0 255 127 255)
        N -> let pos = V2 ((width  `div` 4) * 3)
                          ((height `div` 4) * 3)
                 box1Cycle = rectwaveV2 margins 40 i
                 box2Cycle = rectwaveV2 margins 40 (i + 20)
              in do P.rectangle renderer (pos + box1Cycle - margins - padding) (pos + box1Cycle + nDim + padding) (V4 0 0 255 255)
                    P.rectangle renderer (pos + box2Cycle - margins - padding) (pos + box2Cycle + nDim + padding) (V4 255 0 127 255)

promptLoop :: Renderer  -- renderer to display things onto
           -> F.Font    -- font to display with
           -> String    -- the prompt to display
           -> Selection -- option (either y or n) currently selected
           -> Int       -- iterator
           -> IO Bool   -- true if y is selected, false if n is
promptLoop renderer font prompt sel i =
  do rendererDrawColor renderer $= V4 0 0 0 255
     clear renderer

     renderPrompt renderer font prompt i
     yDim <- renderY renderer font i
     nDim <- renderN renderer font i
     renderSelection renderer sel yDim nDim i

     present renderer

     delay 25

     events <- pollEvents
     let isKeyPress keycode event =
           case eventPayload event of
             KeyboardEvent kbe ->
               keyboardEventKeyMotion kbe == Pressed &&
               keysymKeycode (keyboardEventKeysym kbe) == keycode
             _ -> False
         yPressed     = any (isKeyPress KeycodeY) events
         nPressed     = any (isKeyPress KeycodeN) events
         tabPressed   = any (isKeyPress KeycodeTab) events
         enterPressed = any (isKeyPress KeycodeReturn) events

     -- TODO: add mouse support (click y/n to do what y/n does)
     --       also change selection on hover (i.e. if mouse hovers 
     --       over y, move selection to y)
     -- y means do the command given as argument
     -- n means don't do that command
     if yPressed || enterPressed && sel == Y then pure True
       else if nPressed || enterPressed && sel == N then pure False
       else if tabPressed then (promptLoop renderer font prompt (toggle sel) (i + 1))
       else (promptLoop renderer font prompt sel (i + 1))
